File: Leetcode/NumberofIslands.cpp
class Solution {
public:

vector< vector<int> > used;

void dfs(int i, int j, int n, int m, vector<vector<char>>& board){
	used[i][j] = 1;
	
	if (i - 1 >= 0){
		if (!used[i - 1][j] && board[i - 1][j] == '1'){
			dfs(i - 1, j, n, m, board);
		}		
	}
	if (j - 1 >= 0){
		if (!used[i][j - 1] && board[i][j - 1] == '1'){
			dfs(i, j - 1, n, m, board);
		}
	}

	if (i + 1 < n){
		if (!used[i + 1][j] && board[i + 1][j] == '1'){
			dfs(i + 1, j, n, m, board);
		}
	}

	if (j + 1 < m){
		if (!used[i][j + 1] && board[i][j + 1] == '1'){
			dfs(i, j + 1, n, m, board);
		}
	}
}

int numIslands(vector<vector<char>>& board) {
        int n = board.size();
	    if (n == 0) return 0;
	    int m = board[0].size();
	    vector<int> t; t.assign(m + 1, 0);
	    used.assign(n + 1, t);
	    int ans = 0;
	    for (int i = 0; i < n; ++i){
		    for (int j = 0; j < m; ++j){
			    if (!used[i][j] && board[i][j] == '1'){
			    	ans++;
			    	dfs(i, j, n, m, board);
			    }
		    }
	    }
	    return ans;
    }
};

File: Leetcode/NumberofIslands.rs
// https://leetcode.com/problems/number-of-islands

struct Solution {}

impl Solution {
    const DIRS: &[(i32, i32)] = &[(-1, 0), (1, 0), (0, -1), (0, 1)];

    fn dfs(grid: &mut Vec<Vec<char>>, cur: (usize, usize)) {
        grid[cur.0][cur.1] = 'x';

        let x = Solution::DIRS
            .iter()
            .filter_map(|(di, dj)| {
                let to_i = (cur.0 as i32 + di) as usize;
                let to_j = (cur.1 as i32 + dj) as usize;
                (grid.get(to_i)?.get(to_j)? == &'1').then_some((to_i, to_j))
            })
            .collect::<Vec<_>>();

        for (to_i, to_j) in x.into_iter() {
            if grid[to_i][to_j] != 'x' {
                Solution::dfs(grid, (to_i, to_j));
            }
        }
    }

    pub fn num_islands(mut grid: Vec<Vec<char>>) -> i32 {
        let n = grid.len();
        let m = grid[0].len();

        let mut ans = 0;

        for i in 0..n {
            for j in 0..m {
                if grid[i][j] == '1' {
                    ans += 1;
                    Solution::dfs(&mut grid, (i, j));
                }
            }
        }

        ans
    }
}
